src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -156,47 +156,72 @@ public async Task ExecuteAsync(ExportJobRecord exportJobRecord, WeakETag weakETa
-                return;
+                await HandleJobConflictAsync(cancellationToken);
             }
             catch (Exception ex)
             {
                 // The job has encountered an error it cannot recover from.
                 // Try to update the job to failed state.
                 _logger.LogError(ex, "Encountered an unhandled exception. The job will be marked as failed.");
 
-                await UpdateAndCommitJobStatus(OperationStatus.Failed, updateEndTimestamp: true, cancellationToken);
+                await CompleteJobAsync(OperationStatus.Failed, cancellationToken);
             }
         }
 
-        private async Task UpdateAndCommitJobStatus(OperationStatus operationStatus, bool updateEndTimestamp, CancellationToken cancellationToken)
+        private async Task CompleteJobAsync(OperationStatus completionStatus, CancellationToken cancellationToken)
         {
-            _exportJobRecord.Status = operationStatus;
-
-            if (updateEndTimestamp)
-            {
-                _exportJobRecord.EndTime = Clock.UtcNow;
-            }
+            _exportJobRecord.Status = completionStatus;
+            _exportJobRecord.EndTime = Clock.UtcNow;
 
-            await UpdateAndCommitJobRecord(_exportJobRecord, cancellationToken);
+            await UpdateJobRecordAsync(cancellationToken);
         }
 
-        private async Task UpdateAndCommitJobRecord(ExportJobRecord jobRecord, CancellationToken cancellationToken)
+        private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
         {
             using (IScoped<IFhirOperationDataStore> fhirOperationDataStore = _fhirOperationDataStoreFactory())
             {
-                ExportJobOutcome updatedExportJobOutcome = await fhirOperationDataStore.Value.UpdateExportJobAsync(jobRecord, _weakETag, cancellationToken);
+                ExportJobOutcome updatedExportJobOutcome = await fhirOperationDataStore.Value.UpdateExportJobAsync(_exportJobRecord, _weakETag, cancellationToken);
 
                 _exportJobRecord = updatedExportJobOutcome.JobRecord;
                 _weakETag = updatedExportJobOutcome.ETag;
             }
         }
 
+        private async Task HandleJobConflictAsync(CancellationToken cancellationToken)
+        {
+            // The job was updated by another process.
+            ExportJobOutcome updatedOutcome = null;
+
+            using (IScoped<IFhirOperationDataStore> fhirOperationDataStore = _fhirOperationDataStoreFactory())
+            {
+                updatedOutcome = await fhirOperationDataStore.Value.GetExportJobByIdAsync(_exportJobRecord.Id, cancellationToken);
+            }
+
+            if (updatedOutcome.JobRecord.Status == OperationStatus.Cancelled)
+            {
+                _logger.LogInformation("The job was canceled by another process.");
+
+                // The job was canceled. Merge the output without updating the status.
```

Comment:
Is there a reason we want to store the updated Output? Since the job has anyway been cancelled, will the Output data be used?

Reply:
I think it's still useful to update the Output because consider the following case:

1. Job persist the output information at T1. 
2. Job starts processing the next batches (and let's say `MaximumNumberOfResourcesPerQuery = 100` and `NumberOfPagesPerCommit = 10`)
3. The user cancels the job while Step 2 is happening.

In this case, the job did process additional 1000 resources but updating the job record would fail with JobConflictException because the job was canceled. I think it would be nice to actually update the Output one last time to reflect the latest number otherwise, depending on the configuration values, the discrepancy could be very large.

But this brings up an interesting issue that I didn't think about which is that with the current implementation, it will only stop the job once the job finish processing the batch but the batch size could be large depending on what configuration setting is so it might take a while before the batch finishes. We might need to add check in the loop or something to detect cancellation earlier.

Reply:
I thought about this a bit more and I think you are right. If the job was cancelled before the next commit, there could be additional resources exported but they are not committed yet so it won't show up in the destination anyways so we don't need to update the Output.

Reply:
I agree that it if we are returning the Output to the user, it makes sense that we return the updated one (so that the user can determine what they want to do with it). I initially thought that we will not be returning an Output response for such cancelled export requests. I just read through the spec again and they don't specifically mention how the server should respond (for the get export status call) once the export job gets cancelled. I think we have the freedom to decide whether we want to call it a success or failure.

But marking this as success and returning the Output containing currently exported data might create another kind of issue. Depending on when the client polls for the export status, they can receive different kinds of Output data but the operation status as Cancelled. It might be difficult for the client to determine when they will get the "final" Output data after polling the status. Like you mentioned above, this can be mitigated if we poll for the status within the loop. But is the perf hit that comes with the polling worth it? I feel it might be better to not promise the client anything in the case of a cancelled export job.

Reply:
I added the "lightweight" check using ETag in the loop. It returns 304 Not Modified if the document is not updated and only costs 1 RU as opposed to fetch the whole document. This way it can detect cancellation quicker without putting too much load on the backend. We can try it out and see how it is and decide if we want to keep it or not.

Reply:
This reduces the probability that a user will receive different Output data when polling for the status of a cancelled export job (depending on the timing of the poll and the final job record update) but does not completely eliminate it. I still feel that we are trying to optimize for something that the end-user might not even need (they might just be interested in seeing whether the export job was cancelled rather than what state it was in when it got cancelled).